website/blog/2020-08-20-unnest.md

```diff
@@ -0,0 +1,221 @@
+---
+title: Even Faster Unnest
+author: Ying Su
+authorURL: https://www.linkedin.com/in/ying-su-b00b81107/
+authorFBID: 656599427
+---
+
+Ying Su, Masha Basmanova, Orri Erling
+
+Unnest is a common operation in Facebook’s daily Presto workload. It converts an `ARRAY`, `MAP`, or `ROW` into a flat relation. Its original implementation uses deep copy all the time and was very inefficient. In [Unnest Operator Performance Enhancement with Dictionary Blocks](https://prestosql.io/blog/2019/08/23/unnest-operator-performance-enhancements.html), the author improved the Unnest operator by up to 10x in CPU and elapsed times by using `DictionaryBlock` when possible. We went one step further and improved it for another 5-10x.
+
+<!--truncate-->
+
+ In this post, we will refer to the PrestoSQL implementation as "baseline". The JMH Benchmark results for different cases are shown below ("before" is the baseline implementation, and "after" is our new implementation):
+
+
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestOneColumnWithoutOrdinality.png)
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestTwoColumns.png)
+![Remote Exchange](/img/blog/2020-07-17-unnest.md/UnnestRowType.png)
+
+The optimized Unnest implementation is available in Presto 0.235 and onward and is enabled by default. The JMH benchmark was also expanded to have better coverage and can be found in [BenchmarkUnnestOperator.java](https://github.com/prestodb/presto/blob/29acdbaff128fff2de185c28fb26031be1d88019/presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java).
+
+In Facebook production workload, we observed over a 6x CPU reduction on this operator. The following chart shows the pariwise comparison for all unnest operatrs in one of our test runs. Every dot below the diagonal is a win. About 25% of operators show over a 5x CPU reduction, and some of them have even over a 20x reduction.
```

Comment:
```suggestion
In Facebook's production workload, we observed over a 6x CPU reduction on this operator. The following chart shows the pairwise comparison for all unnest operators in one of our test runs. Every dot below the diagonal is a win. About 25% of operators show over a 5x CPU reduction, and some of them have even over a 20x reduction.
```

Reply:
@aweisberg Ariel thank you so much! I have corrected these typos and moved the charts to `/img/blog/2020-08-20-unnest.md`. Will you be able to merge it?